sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,285 @@
+      // in the batch. This is acceptable because batching is a best mechanism
+      // effort here.
+      do
+      {
+        cv_.wait_for(lk, std::chrono::milliseconds(timeout));
+
+        // If shutdown has been invoked, break out of the loop
+        // and drain out the queue.
+        if (is_shutdown_ == true)
+        {
+          is_shutdown_ = false;
+          DrainQueue();
+          return;
+        }
+
+        if (is_force_flush_ == true)
+        {
+          break;
+        }
+      } while (buffer_->empty() == true);
+    }
+
+    // Get the value of is_force_flush_ to check whether this export is
+    // a result of a ForceFlush call. This flag is propagated to the Export
+    // method as well.
+    bool was_force_flush_called = is_force_flush_;
+
+    // Based on whether or not ForceFlush was invoked, consume and copy
+    // the right amount of spans from the buffer_ into another buffer.
+    auto buffer_copy = CopySpans(was_force_flush_called);
+
+    // Set the flag back to false to notify main thread in case ForceFlush
+    // was invoked.
+    if (is_force_flush_ == true)
+    {
+      is_force_flush_ = false;
+    }
+
+    // Unlock here to unblock all producers
+    lk.unlock();
+
+    auto start = std::chrono::steady_clock::now();
+    Export(buffer_copy, was_force_flush_called);
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    timeout = schedule_delay_millis_ - duration;
+  }
+}
+
+void BatchSpanProcessor::Export(std::unique_ptr<common::CircularBuffer<Recordable>> &buffer,
+                                const bool was_force_flush_called)
+{
+  std::vector<std::unique_ptr<Recordable>> spans_arr;
+
+  buffer->Consume(
+      buffer->size(), [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+          std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+          ptr.Swap(swap_ptr);
+          spans_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));
+
+  if (is_shutdown_ == true)
+  {
+    return;
+  }
+
+  // Notify the main thread in case this export was the result of a force flush.
+  if (was_force_flush_called == true)
+  {
+    is_force_flush_          = false;
+    is_force_flush_notified_ = true;
+
+    while (is_force_flush_notified_ == true)
+    {
+      force_flush_cv_.notify_one();
+    }
+  }
+}
+
+std::unique_ptr<common::CircularBuffer<Recordable>> BatchSpanProcessor::CopySpans(
+    const bool was_force_flush_called)
+{
+  // Make a copy of the buffer to avoid blocking other producers.
+  std::unique_ptr<common::CircularBuffer<Recordable>> buffer_copy(
+      new common::CircularBuffer<Recordable>(max_queue_size_));
```

Comment:
That shouldn't be necessary. `CircularBuffer` should be able to deal with concurrent reads/writes.

Reply:
I believe I can get rid of this entire function now.

Reply:
The purpose of this function was to record the exact state of the buffer at a given point in time. 

Let's say ForceFlush was invoked, which in turned required that the entire buffer be flushed out. In that case, if we don't record/duplicate the exact state of the buffer when the ForceFlush method was invoked, it's possible that other producers keep adding spans to the buffer while the buffer is being consumed.

Moreover, recording the state of the buffer helps empty the buffer for other spans. If not emptied out, it's possible that some spans, which shouldn't have been dropped, get dropped.